### ISSUE TITLE
[rancher-dns] Illegal UDP message size breaks name resolution for container applications

### ISSUE THREAD

**Issue by janeczku** [rancher/rancher#2928]
Rancher DNS server may overflow the UDP message size limit of 512 bytes in replies sent to a client when doing a recursive query.
Sending messages larger than 512 bytes over UDP is technically illegal (see RFC1035) and breaks name resolution for DNS clients that do not support EDNS0. The correct behavior would be to set the Truncated bit on these messages.
For example Golang net resolver can't deal with packets larger than 512 bytes: The lookup then fails with a generic error: "no such host"

Example: 
The host has an upstream DNS server that includes additional section and extra section records.
Starting the Rancher Route53 service fails:

```TEXT
/ # /usr/bin/external-dns -debug
INFO[0000] CLOUDFLARE_EMAIL is not set, skipping init of CloudFlare provider 
INFO[0000] DNSIMPLE_TOKEN is not set, skipping init of DNSimple provider 
FATA[0000] Failed to list hosted zones: Get https://route53.amazonaws.com/2013-04-01/hostedzone/?maxItems=9223372036854775807:
dial tcp: lookup route53.amazonaws.com on 169.254.169.250:53: no such host 
```

Here is the proof that Rancher DNS overflows the UDP packet size limit. There is also another quirk:  The message has almost exactly double the byte size than when querying the upstream server directly. Some kind of duplication or missing compression?
*Note: The reply has neither Truncated flag set nor does it have the EDNS0 extension. Even though 'dig' tool is still able to parse the message, but that doesn't mean that the reply is not totally illegal!*

```TEXT
/ # dig A api.cloudflare.com
; <<>> DiG 9.10.2-P4 <<>> A api.cloudflare.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 13973
;; flags: qr rd ra; QUERY: 1, ANSWER: 6, AUTHORITY: 5, ADDITIONAL: 10

;; QUESTION SECTION:
;api.cloudflare.com.		IN	A

;; ANSWER SECTION:
api.cloudflare.com.	70	IN	CNAME	api-origin.cloudflare.com.
api-origin.cloudflare.com. 70	IN	A	198.41.215.67
api-origin.cloudflare.com. 70	IN	A	198.41.215.68
api-origin.cloudflare.com. 70	IN	A	198.41.214.67
api-origin.cloudflare.com. 70	IN	A	198.41.214.68
api-origin.cloudflare.com. 70	IN	A	198.41.215.66

;; AUTHORITY SECTION:
cloudflare.com.		92749	IN	NS	ns4.cloudflare.com.
cloudflare.com.		92749	IN	NS	ns5.cloudflare.com.
cloudflare.com.		92749	IN	NS	ns6.cloudflare.com.
cloudflare.com.		92749	IN	NS	ns7.cloudflare.com.
cloudflare.com.		92749	IN	NS	ns3.cloudflare.com.

;; ADDITIONAL SECTION:
ns3.cloudflare.com.	132654	IN	A	162.159.0.33
ns3.cloudflare.com.	28829	IN	AAAA	2400:cb00:2049:1::a29f:21
ns4.cloudflare.com.	132654	IN	A	162.159.1.33
ns4.cloudflare.com.	28829	IN	AAAA	2400:cb00:2049:1::a29f:121
ns5.cloudflare.com.	132654	IN	A	162.159.2.9
ns5.cloudflare.com.	28829	IN	AAAA	2400:cb00:2049:1::a29f:209
ns6.cloudflare.com.	132654	IN	A	162.159.3.11
ns6.cloudflare.com.	7283	IN	AAAA	2400:cb00:2049:1::a29f:30b
ns7.cloudflare.com.	132654	IN	A	162.159.4.8
ns7.cloudflare.com.	132654	IN	AAAA	2400:cb00:2049:1::a29f:408

;; Query time: 12 msec
;; SERVER: 169.254.169.250#53(169.254.169.250)     <------ Rancher DNS server
;; WHEN: Sat Dec 05 07:35:36 UTC 2015
;; MSG SIZE  rcvd: 928                                <------- 928 bytes - Overflow!
```

When querying the upstream server directly we get exactly the same records but with half the byte size and conforming to the 512 bytes limit.
```TEXT
/etc # dig A api.cloudflare.com @213.186.33.99

; <<>> DiG 9.10.2-P4 <<>> A api.cloudflare.com @213.186.33.99
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 27269
;; flags: qr rd ra; QUERY: 1, ANSWER: 6, AUTHORITY: 5, ADDITIONAL: 11

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;api.cloudflare.com.		IN	A

;; ANSWER SECTION:
api.cloudflare.com.	70	IN	CNAME	api-origin.cloudflare.com.
api-origin.cloudflare.com. 70	IN	A	198.41.215.68
api-origin.cloudflare.com. 70	IN	A	198.41.214.67
api-origin.cloudflare.com. 70	IN	A	198.41.214.68
api-origin.cloudflare.com. 70	IN	A	198.41.215.66
api-origin.cloudflare.com. 70	IN	A	198.41.215.67

;; AUTHORITY SECTION:
cloudflare.com.		92448	IN	NS	ns3.cloudflare.com.
cloudflare.com.		92448	IN	NS	ns4.cloudflare.com.
cloudflare.com.		92448	IN	NS	ns7.cloudflare.com.
cloudflare.com.		92448	IN	NS	ns6.cloudflare.com.
cloudflare.com.		92448	IN	NS	ns5.cloudflare.com.

;; ADDITIONAL SECTION:
ns3.cloudflare.com.	132353	IN	A	162.159.0.33
ns3.cloudflare.com.	28528	IN	AAAA	2400:cb00:2049:1::a29f:21
ns4.cloudflare.com.	132353	IN	A	162.159.1.33
ns4.cloudflare.com.	28528	IN	AAAA	2400:cb00:2049:1::a29f:121
ns5.cloudflare.com.	132353	IN	A	162.159.2.9
ns5.cloudflare.com.	28528	IN	AAAA	2400:cb00:2049:1::a29f:209
ns6.cloudflare.com.	132353	IN	A	162.159.3.11
ns6.cloudflare.com.	6982	IN	AAAA	2400:cb00:2049:1::a29f:30b
ns7.cloudflare.com.	132353	IN	A	162.159.4.8
ns7.cloudflare.com.	132353	IN	AAAA	2400:cb00:2049:1::a29f:408

;; Query time: 11 msec
;; SERVER: 213.186.33.99#53(213.186.33.99)
;; WHEN: Sat Dec 05 07:40:37 UTC 2015
;; MSG SIZE  rcvd: 462                                    <------- Half the size!
```

**Comment by vincent99** [comment]
This sounds like https://github.com/miekg/dns/issues/216 .  Pretty much the entirety of requests to a recursive server is calling `Exchange()`.  All the extra copies of the labels are 400+ bytes on that record when sent without compression.

**Comment by janeczku** [comment]
@vincent99 That explains the larger size indeed. The problem here is that Rancher DNS just blindly passes the recursive response on to the client without making sure it conforms to the 512-byte UDP size restriction.

**Comment by janeczku** [comment]
# Steps to reproduce

1. Change the nameserver in one of the host's /etc/resolv.conf to 194.120.55.3 (This an example of a nameserver that includes additional/extra sections in the answers).
2. Restart the Network Agent on that host so that Rancher DNS uses that server as upstream recursor.
3. Deploy a new container to that host using the [janeczku/dnstest-alpine](https://github.com/janeczku/dnstest) image with network option set to 'managed'. This will run a Go application that does nothing but resolve some domain names using `net.LookupHost()` calls and print the results.
4. When the container has started check the logs -> the test app fails to resolve domains

**Comment by janeczku** [comment]
@vincent99 I guess this can be closed since https://github.com/rancher/rancher-dns/pull/9 has been merged?

**Comment by will-chan** [comment]
@janeczku did you try this out?  If it works for you, let's close this issue.

**Comment by janeczku** [comment]
@will-chan 
Confirmed fixed in Rancher v0.50.0. DNS packet returned from Rancher DNS is now within the legal limit of 512 bytes.

```TEXT
$ dig -t A api.cloudflare.com

; <<>> DiG 9.9.5-3ubuntu0.2-Ubuntu <<>> A api.cloudflare.com          
;; global options: +cmd                                                         
;; Got answer:                                                                  
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 3698                        
;; flags: qr rd ra ad; QUERY: 1, ANSWER: 6, AUTHORITY: 5, ADDITIONAL: 10 

(...)

;; Query time: 12 msec                                                          
;; SERVER: 169.254.169.250#53(169.254.169.250)                                  
;; WHEN: Tue Dec 22 21:17:08 UTC 2015                                           
;; MSG SIZE  rcvd: 451
```

**Comment by will-chan** [comment]
@janeczku awesome!  Closing issue.
